File: node/src/routes/replays.ts
import { Router } from "express";
import { z } from "zod";
import { pool } from "../db/pool.js";
import { asyncHandler } from "../lib/asyncHandler.js";

export const replaysRouter = Router();

const replaySchema = z
  .object({
    range_start: z.string().datetime(),
    range_end: z.string().datetime(),
  })
  .refine((body) => new Date(body.range_end) >= new Date(body.range_start), {
    message: "range_end must not be before range_start",
  });

interface ReplayRow {
  id: string;
  status: string;
}

replaysRouter.post(
  "/endpoints/:id/replays",
  asyncHandler(async (req, res) => {
    const parsed = replaySchema.safeParse(req.body);
    if (!parsed.success) {
      res.status(400).json({ error: { code: "invalid_request", message: parsed.error.issues[0]?.message ?? "Invalid request body" } });
      return;
    }

    const idempotencyKey = req.header("idempotency-key");
    if (!idempotencyKey) {
      res.status(400).json({ error: { code: "invalid_request", message: "Idempotency-Key header required" } });
      return;
    }

    const { rows: endpointRows } = await pool.query(`SELECT id FROM endpoints WHERE id = $1 AND tenant_id = $2`, [
      req.params.id,
      req.tenantId,
    ]);
    if (!endpointRows[0]) {
      res.status(404).json({ error: { code: "not_found", message: "Endpoint not found" } });
      return;
    }

    const { range_start, range_end } = parsed.data;
    const inserted = await pool.query<ReplayRow>(
      `INSERT INTO replays (endpoint_id, idempotency_key, range_start, range_end)
       VALUES ($1, $2, $3, $4)
       ON CONFLICT (endpoint_id, idempotency_key) DO NOTHING
       RETURNING id, status`,
      [req.params.id, idempotencyKey, range_start, range_end],
    );
    let replay = inserted.rows[0];
    if (!replay) {
      const existing = await pool.query<ReplayRow>(`SELECT id, status FROM replays WHERE endpoint_id = $1 AND idempotency_key = $2`, [
        req.params.id,
        idempotencyKey,
      ]);
      replay = existing.rows[0]!;
    }

    res.status(202).json({ id: replay.id, status: replay.status });
  }),
);

File: go/internal/api/replays.go
package api

import (
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/jackc/pgx/v5"

	"webhooks-go/internal/auth"
	"webhooks-go/internal/httpx"
)

type replayRequest struct {
	RangeStart string `json:"range_start"`
	RangeEnd   string `json:"range_end"`
}

// docs/adr/0005: replay mirrors publish's async two-phase shape exactly —
// this handler does one thing synchronously (insert the replays row) and
// returns 202 immediately; the durable ack is that one insert. The shared
// worker pool later expands a pending replay (internal/worker's
// RunReplayExpansionCycle).
func (s *Server) createReplay(w http.ResponseWriter, r *http.Request) {
	idempotencyKey := r.Header.Get("Idempotency-Key")
	if idempotencyKey == "" {
		httpx.WriteError(w, http.StatusBadRequest, "invalid_request", "Idempotency-Key header required")
		return
	}

	var req replayRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		httpx.WriteError(w, http.StatusBadRequest, "invalid_request", "Invalid request body")
		return
	}
	rangeStart, err := time.Parse(time.RFC3339, req.RangeStart)
	if err != nil {
		httpx.WriteError(w, http.StatusBadRequest, "invalid_request", "range_start must be an RFC3339 datetime")
		return
	}
	rangeEnd, err := time.Parse(time.RFC3339, req.RangeEnd)
	if err != nil {
		httpx.WriteError(w, http.StatusBadRequest, "invalid_request", "range_end must be an RFC3339 datetime")
		return
	}
	if rangeEnd.Before(rangeStart) {
		httpx.WriteError(w, http.StatusBadRequest, "invalid_request", "range_end must not be before range_start")
		return
	}

	endpointID := r.PathValue("id")
	var exists string
	err = s.pool.QueryRow(r.Context(),
		"SELECT id FROM endpoints WHERE id = $1 AND tenant_id = $2", endpointID, auth.TenantID(r),
	).Scan(&exists)
	if fail(w, "createReplay: endpoint lookup", err, "Endpoint not found") {
		return
	}

	// ON CONFLICT DO NOTHING returns zero rows on a conflict — the repeated
	// call must return the *original* replay's id/status, so a conflict
	// needs a follow-up SELECT rather than trusting the INSERT's RETURNING
	// (same idempotency shape as POST /events).
	var id, status string
	err = s.pool.QueryRow(r.Context(),
		`INSERT INTO replays (endpoint_id, idempotency_key, range_start, range_end)
		 VALUES ($1, $2, $3, $4)
		 ON CONFLICT (endpoint_id, idempotency_key) DO NOTHING
		 RETURNING id, status`,
		endpointID, idempotencyKey, rangeStart, rangeEnd,
	).Scan(&id, &status)
	if errors.Is(err, pgx.ErrNoRows) {
		err = s.pool.QueryRow(r.Context(),
			`SELECT id, status FROM replays WHERE endpoint_id = $1 AND idempotency_key = $2`,
			endpointID, idempotencyKey,
		).Scan(&id, &status)
	}
	if fail(w, "createReplay", err, "") {
		return
	}

	httpx.WriteJSON(w, http.StatusAccepted, struct {
		ID     string `json:"id"`
		Status string `json:"status"`
	}{id, status})
}
